README.md
<p align="center"><a href="https://devui.design/home" target="_blank" rel="noopener noreferrer"><img alt="DevUI Logo" src="logo.svg?sanitize=true" width="180" style="max-width:100%;">
</p>
<p align="center">
  <a href="https://github.com/DevCloudFE/DevUIHelper"><img src="https://img.shields.io/github/stars/DevCloudFE/DevUIHelper.svg?label=github%20stars" alt="Github Star"></a>
</br>
  <a href="README.md"><img src="https://img.shields.io/badge/document-English-blue" alt="Document"></a>
  <a href="README_zh_CN.md"><img src="https://img.shields.io/badge/%E6%96%87%E6%A1%A3-%E4%B8%AD%E6%96%87-blue" alt="Document"></a>
  <a href="https://marketplace.visualstudio.com/items?itemName=sspkuDevUI.devuihelper"><img src="https://img.shields.io/visual-studio-marketplace/v/sspkuDevUI.devuihelper" alt="version"></a>
  <a href="https://marketplace.visualstudio.com/items?itemName=sspkuDevUI.devuihelper"><img src="https://img.shields.io/visual-studio-marketplace/i/sspkuDevUI.devuihelper" alt="installs"></a>
</p>

<h1 align="center">DevUIHelper for DevUI Component</h1>


## Plug-in Features

### Completing Tips
Provides the automatic code completion function for components and instructions, automatically supplements mandatory parameters, and provides information about all APIs supported by components and instructions for selection.
![completion](completion-example.gif)

### Hover Information 
Hovers show information about the symbol/object that's below the mouse cursor, including application scenarios, supported API lists, and API details.
![hover](hover-tips-example.gif)

### Code error correction
- Alerts when devui content is used and devui is not installed.
- Alarm if no necessary entry is made under the devui tab.

## Development and Debugging
- Run `npm install` in the root directory.
- Run `npm run compile` to compile client and server
- Switches to the debug view.
- Select `Launch Client` from the drop-down list.
- Run the startup configuration.
- If you also want to debug the server, use the startup configuration `Attach to Server`


## Contribution

You are welcome to contribute your code or discuss your great ideas!

Thanks to all the contributors who helped us build DevUI


CHANGELOG.md
<!--
 * @Author: your name
 * @Date: 2020-03-14 17:21:08
 * @LastEditTime: 2020-06-08 14:59:50
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \DevUI-Language-Support\CHANGELOG.md
 -->

# 2020/7/9 V4.4.1
- 增加了tag必要属性的检测。
- 增加了错误屏蔽，现在当您的标签有错误提醒时，将无法使用悬停提示，直到您解决了错误。
- 属性的提示文本中添加了属性来源。

# 2020/7/3 V4.4.0
- 增加了devui安装提示。
- 更新了悬浮提示与补全的内容样式。
- 支持自定义插件功能，用户可以选择关闭或者开启悬浮提示。

# 2020/6/24 V4.3.9
- 支持devui开发者模式
- 修复冗余提示的bug
- 更新Readme,提供最新版功能的介绍

# 2020/6/22 V4.3.4
- 表达式支持{}语法运算，现在可以指定标签内部的内容
- 修复表达式与解析器的边界问题

# 2020/6/21 V4.3.3
- 修改表达式解析器为递归式，允许解析更为复杂的表达式
- 支持#语法，现在能输入id了

# 2020/6/16 V4.3.2
- 将表达式解析器替换为Cursor驱动
- 扩充表达式语法，支持[]运算

# 2020/6/13 V4.3.1
- 修复表达式引起的循环查找BUG

# 2020/6/13 V4.3.0
- @表达式支持，在标签内容中尝试着输入@d.but.pri试试
- 修复了普通标签不能触发指令的bug

# 2020/6/8 V4.2.5
- 支持指令类型的悬停提示

# 2020/6/6 V4.2.2 
- 添加了由于将资源树迁移至structure类导致的一系列bug
- 更新wchc脚本

# 2020/4/13 V3.0.0
- 定制化parser，基于语法树为用户提供快速提示功能

# V2.3.0
- 为自动补全提供了新的外观
- 现在我们为属性提供了补全服务，并且让它能够正确的停在两个引号之间
- 修复了多次换行导致补全与悬浮提示失效的问题
- 修复了多个空格引发的补全不稳定问题
- 修复了标签`d-column`不存在的问题



docs-scripts/attrListScanner.js
/*
 * @Author: your name
 * @Date: 2020-06-04 16:31:48
 * @LastEditTime: 2020-06-05 16:27:48
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \Extension_Universe\WCH-Creater\attrListCreater.js
 */ 
const util = require('./util');
const creater= require('./creater');
const fs = require('fs');
let componentInbuild;
let count = 0;

/**
 * api处理大师
 */
const _apireader = (arr,i,flag)=>{
    let result = creater.result;
    //初始化
    let componentName = util.getComponentName(arr[i]);
    const directiveFlag = componentName.charAt(1)!=="-";
    result[componentName] = result[componentName]?result[componentName]:{name:componentName,attrList:[],directiveFlag:directiveFlag};
    componentattrs=result[componentName]['attrList']?result[componentName]['attrList']:[];

    //过滤表格
    while(arr[++i]&&arr[i].indexOf(':-')===-1){
        if(!arr[i].startsWith('\|')&&arr[i].match(/(\S{5,})/)){
            result[componentName]['description'] = arr[i].replace(/(\s|\#|\(|\))/g,'')
        }
        if(arr[i].match(/(\#)*(\s)*d/)){
            break;
        }
    }
    i++;

    //装载
    while(arr[i]&&arr[i].startsWith('\|')){
        let temp = arr[i];
        temp = temp.replace(/\\\|/g,"!!!")
        let temparr = temp.split('\|');
        if(flag){
            if(temparr.length!==6){
                console.log("Error in"+temp)
                console.log(",this problem may raised by lack of | in the end of line or cotains \'\|\' rather than\'\\\|\' in discription " )
            }
            temparr = util.washContent(temparr);
            const typeResult = util.typeFactory(temparr[1])
            componentattrs.push({
                name:temparr[0],
                type:typeResult[0],
                default:temparr[2],
                description:temparr[3],
                attrType:'ATTR',
                isNecessary:temparr[2].startsWith('必选'),
                valueSet:typeResult[1]?typeResult[1]:[],
                isEvent:false
            })
            // console.log(temparr);
        }
        //对事件进行操作
        else{
            if(temparr.length!==5){
                console.log("Error in"+temp)
                console.log(",this problem may raised by lack of | in the end of line or cotains \'\|\' rather than\'\\\|\' in discription " )
            }
            temparr = util.washContent(temparr);
            const type = util.typeFactory(temparr[1])
            componentattrs.push({
                name:temparr[0].replace(/\(|\)/g,""),
                type:type,
                description:temparr[2],
                attrType:'Event',
                isNecessary:temparr[2].startsWith('必选'),
                hasvalueSet:type instanceof Array,
                isEvent:true
            })    
        }
        // result[componentName].attrlist=componentattrs;
        i++;
    }
    
}
let i=0;
const apireader = (content)=>{

    let arr = content.split('\n');
    for (let index = 0; index < arr.length; index++) {
        let temp = arr[index];
        if(blackListCheck(temp)){
            continue;
        }
        //处理括号的注释
        let tempStringArr = temp.split(" ")
        if(tempStringArr.length===3&&tempStringArr.pop().match(/\((\S*)\)/)){
            temp = tempStringArr[1];
        }
        else if(temp.match((/参数(\s*)(\r*)$/))){
            console.log(i+++temp)
            _apireader(arr,index,true);
        }else if(temp.match(/事件(\s*)(\r*)$/)){
            console.log(i+++temp)
            _apireader(arr,index,false);
        }else if(temp.match(/指令(\s*)(\r*)$/)){
            console.log(i+++temp)
            _apireader(arr,index,true);
        }
    }
}
attrListScanner = (rootPath)=>{
    return new Promise((resolve,reject)=>{
        fs.readdir(rootPath+'/devui',(err,files)=>{
            if(err)
            reject(err.message)
            else{
                files.forEach(file=>{
                    fs.readFile(rootPath+'/devui/'+file+'/doc/api.md',(err,data)=>{
                        if(err){
                            // console.log(`this file ${file} does not have api`)
                        }else{
                            apireader(data.toString());
                        }
                    });
                })
                resolve('Done');
            }
        })
    });
    await 
}
blackListCheck=(s)=>{
    const blackList = ["判断激活锚点的事件",
    "上传额外自定义参数","使用 `dDraggable` & `dDroppable` 指令",
    "Image容器元素上使用`dImagePreview`指令",
    "提供dDraggable和dDroppable(dSortable)指令",
    " 模板参数",
    "IDialogOptions 事件"];
    let res = false;
    blackList.forEach((e)=>{
        if(s.indexOf(e)!==-1){
            res = true;
        }
    })
    return res;
}
module.exports.attrListScanner =attrListScanner;

docs-scripts/componentScanner.js
/*
 * @Author: your name
 * @Date: 2020-06-04 16:56:19
 * @LastEditTime: 2020-06-04 19:18:12
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \Extension_Universe\WCH-Creater\ComponentScanner.js
 */ 
const fs = require('fs'); 
const creater = require('./creater');
const ComponentScanner = async (rootPath)=>{
    return new Promise((resolve,reject)=>{
        let result = creater.result;
        fs.readFile(rootPath+'/src/app/component/component.route.ts',(err,data)=>{
            if(err){
                reject(err.message);
            }else{
                let dataString = data.toString();
                dataString = dataString.substring(dataString.indexOf('=')+1);
                dataString = dataString.replace(/\r|\n/g," ");
                dataString = dataString.replace(/\'|\`/g,"\"");
                dataString= dataString.replace(/(?:\s*['"]*)?([a-zA-Z0-9]+)(?:['"]*\s*)?:/g, "\"$1\":");
                dataString= dataString.replace(/\"component\"\:(\s\S*)\,/g, "");
                dataString= dataString.replace(/\"loadChildren\"\:(\s|\S)*?\,/g, "");
                dataString= dataString.replace(/<(\/)*[a-zA-z]*>/g, "");
                dataString= dataString.replace(/\}\,(\s*)\]\;(\s*)$/,"}]")
     
                let info = JSON.parse(dataString);
                // console.log(info);
                info.forEach((e)=>{
                    for(let componentName in result ){
                        // console.log(componentName);
                        if(e['path']&&componentName.endsWith(e['path'])){
                            result[componentName]['description']=e.data.description;
                            result[componentName]['tmw'] = e.data.tmw;
                            result[componentName]['cnName'] = e.data.cnName;
                        }
                    }
                })
                resolve('Done');
            }
        })
    });
   
}
module.exports.ComponentScanner = ComponentScanner;
           

docs-scripts/creater.js
/*
 * @Author: your name
 * @Date: 2020-06-03 18:29:27
 * @LastEditTime: 2020-06-05 17:59:06
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \Extension_Universe\WCH-Creater\creater.js
 */ 
const fs = require('fs');
// const rootFile = require('C:\\Users\\Administrator\\Documents\\GitHub\\angular\\ng-devui/src/app/component/component.route.ts');
// const schema = rootFile.routesConfig;
const util = require('./util');
const apic = require('./attrListScanner');
const ComponentScanner = require('./componentScanner');
let result = {};
let _result = [];
//true表示参数

const creater =  async function(rootPath){
    console.log('creater works!');

    await apic.attrListScanner(rootPath);
    await ComponentScanner.ComponentScanner(rootPath);
    for(let c in result){
        _result.push(result[c]);
    }
    fs.mkdir(rootPath+'/wch',(err)=>{
        console.log(err);
    });
    fs.writeFile(rootPath+'/wch/info.json',JSON.stringify(_result),err=>{
        if(err){
            console.log(err);
        }
        console.log('Done!')
    });

}
module.exports.result=result;
module.exports.creater=creater;

creater();